Add security section (#3853)
fleetdm/fleet: handbook/security.md

* [Gatekeeper](https://support.apple.com/en-ca/HT202491) is a feature on macOS that verifies if applications are properly signed by the developer, and notarized by Apple, a process where they do some testing on the application before "stamping" it. The certificates used by applications can be revoked, for example, if a vendor is discovered to be bundling malware in legitimate applications. With Gatekeeper enabled, unsigned and/or unnotarized applications will not be executed with the standard double-click of the icon. This is a useful first line of defense to have.
* Using the firewall will ensure that we limit the exposure to our computers, since they are on untrusted networks. Stealth mode makes them more difficult to discover. 
* Firewall logging allows us to troubleshoot and investigate, by letting us know if applications or connections are being blocked by it.

**User experience impact**

> * A password will be needed as soon as the laptop is turned on, due to the encryption process, instead of once the laptop has booted.

Author: Exactly!

* No performance impact - macOS encrypts the system drive by default. 
* Unsigned or unnotarized applications will require extra steps to execute.
* Unsigned applications will not be allowed to open a firewall port for inbound connections.

#### Screen saver and automatic locking

| #     | Setting                                                                             |
| ----- | ----------------------------------------------------------------------------------- |

> | 2.3.1 | Ensure an inactivity interval of 20 minutes or less for the screen saver is enabled |

Author: The text in these tables is straight from the CIS requirements, I'd rather not edit it.

> | 6.1.2 | Ensure show password hints is Disabled                                              |

Author: The text in these tables is straight from the CIS requirements, I'd rather not edit it.

> | 6.1.2 | Ensure show password hints is Disabled                                              |

Author: But uppercase/lower case makes sense!

| 6.1.3 | Ensure guest account is disabled                                                    |

> | NA    | Prevent the use of automatic logon                                                  |

Author: The text in these tables is straight from the CIS requirements, I'd rather not edit it.

> | NA    | Prevent the use of automatic logon                                                  |

Reviewer: Actually, perhaps "logon" is the correct usage after all. After some digging, it seems like "logon" is appropriate for hardware/devices, while "login" is more for software, so I'm good to keep it as "logon."

I'm less thrilled though about migrating third party's typos and grammar inconsistencies into the handbook e.g. the inconsistencies with capitalization in some of the CIS table entries.

> | NA    | Prevent the use of automatic logon                                                  |

Author: Yeah let's fix capitalization, but the actual wording... unless it's super egregious I would prefer to keep.

**Why?**

* Our workstations are laptops, which by definition are portable and used in many different areas. If a laptop is forgotten, while logged in, it defeats the purpose of other controls and exposes data that could be critical. 
* Password hints can be dangerous, as they can sometimes be easier to guess than the password itself. Since we can support employees remotely via MDM, and since we do not require passwords that expire, we eliminate that risk by hiding the hints.
* Guest accounts are not useful for systems that are used by a single employee, therefore, we disable them.

> * Automatic logon would defeat the purpose of even requiring passwords to unlock computers.

Author: No, automatic logon isn't necessarily *guest* - you can configure your Mac to log in as you automatically! (EEP! - though in reality with Filevault it wouldn't work anyway)

**User experience impact**

* Laptops will lock after 20 minutes of inactivity. To voluntarily pause this, you can configure a [hot corner](https://support.apple.com/en-mo/guide/mac-help/mchlp3000/mac) to disable the screen saver. This can be useful if you are, for example, watching an online meeting without moving the mouse and want to be sure the laptop will not lock.
* Forgotten passwords will have to be fixed via MDM, instead of relying on potentially dangerous hints.
* Guest accounts will not be available.

#### iCloud

> We do not apply ultra restrictive [Data Loss Prevention](https://en.wikipedia.org/wiki/Data_loss_prevention_software) style policies to our workstations. A computer that is used for day to day work, with full access to the Internet can never be protected from voluntary malicious actors willing to upload data. Instead, we focus on ensuring the most critical data never reaches our laptops, so it can remain secure, while our laptops can remain productive.

Author: More like, malware and people controlling the malware using your laptop to steal data. But it could also be a malicious employee - for example in banks they usually block USB devices completely because of their threat model BUT there always are so many other ways to exfiltrate data that it's usually just annoying and does not even improve security much when companies do that.

> We do not apply ultra restrictive [Data Loss Prevention](https://en.wikipedia.org/wiki/Data_loss_prevention_software) style policies to our workstations. A computer that is used for day to day work, with full access to the Internet can never be protected from voluntary malicious actors willing to upload data. Instead, we focus on ensuring the most critical data never reaches our laptops, so it can remain secure, while our laptops can remain productive.

Reviewer: Gotcha regarding malicious actors, but I'm wondering whether we need to include it. It's quite full-on, and hard for me to digest as an end-user/employee who just navigated here to check out our policy on iCloud usage.

I think this is easier to digest and more consistent with the other profile settings throughout the doc (i.e this is the only settings table with an intro paragraph)

> We do not apply ultra restrictive [Data Loss Prevention](https://en.wikipedia.org/wiki/Data_loss_prevention_software) style policies to our devices. Instead, we focus on ensuring the most critical data never reaches our laptops, so it can remain secure, while our laptops can remain productive.

Can we use that instead?

> We do not apply ultra restrictive [Data Loss Prevention](https://en.wikipedia.org/wiki/Data_loss_prevention_software) style policies to our workstations. A computer that is used for day to day work, with full access to the Internet can never be protected from voluntary malicious actors willing to upload data. Instead, we focus on ensuring the most critical data never reaches our laptops, so it can remain secure, while our laptops can remain productive.

Reviewer: Even that's a bit confusing though. "What the heck is the critical data, and why does it never reach my laptop?" thought Bob from HR.

How about:

> We do not apply ultra restrictive Data Loss Prevention style policies to our devices. Instead, by using our company Google Drive, we ensure that the most critical company data never reaches our laptops, so it can remain secure, while our laptops can remain productive.

Then eventually link Google Drive to its relevant section of the handbook when it's been added.

(Assuming that I've understood all this correctly.)

**Why?**
* We do not use managed Apple IDs, and allow employees to use their own iCloud accounts. We disable this to avoid "accidental" copying of data to iCloud, but still allow iCloud drive.

**User experience impact**

* iCloud remains allowed, but the Desktop and Documents folders will not be synchronized. Ensure you put your documents in our Google Drive, so you do not lose them if your laptop has an issue.

#### Miscellaneous security settings

| #     | Setting                                                      |
| ----- | ------------------------------------------------------------ |
| 2.5.6 | Ensure limit ad tracking is enabled                          |
| 2.10  | Ensure secure keyboard entry terminal.app is enabled         |
| 5.1.4 | Ensure library validation is enabled                         |
| 6.3   | Ensure automatic opening of safe files in Safari is disabled |

**Why?**

* Limiting ad tracking has privacy benefits, and no downside.
* Protecting keyboard entry into the terminal app could prevent malicious applications, or non-malicious but inappropriate applications from receiving passwords.

> * Library validation ensures that applications can't be tricked into loading a library in a different location, leaving it open to abuse.

Author: Malware or an attacker exploiting a vulnerability can hijack execution by tricking an application into loading a malicious software library

> * Library validation ensures that applications can't be tricked into loading a library in a different location, leaving it open to abuse.

Reviewer: Aha, understood. What do you think about:

> Library validation ensures that an attacker can't trick applications into loading a software library in a different location, leaving it open to abuse.

Adding the subject ("an attacker") and changing "library" to "software library" helps me understand the context better.

> * Library validation ensures that applications can't be tricked into loading a library in a different location, leaving it open to abuse.

Author: Applying that!

* Safari opening files automatically can lead to negative scenarios where files are downloaded and automatically opened in another application. Though the setting relates to files deemed "safe", it includes PDFs and other file formats where malicious documents exploiting vulnerabilities have been seen before.

**User experience impact**

* Minimal to invisible for these settings, however, applications used to create custom keyboard macros will not be able to receive keystrokes when Terminal.app is the active application window.

#### Enforce DNS-over-HTTPs (DoH)

| #  | Setting                |
| -- | ---------------------- |
| NA | Enforce [DNS over HTTPS](https://en.wikipedia.org/wiki/DNS_over_HTTPS) |

**Why?**

> * We assume laptops are used on dangerous networks. Therefore, DNS queries could be exposed, and leak private data. An attacker on the same wireless network could see DNS queries, determine who your employer is, or even intercept them and [respond with malicious answers](https://github.com/iphelix/dnschef). Using DoH protects the DNS queries from eavesdropping and tampering.

Reviewer: ```suggestion
* We assume that no network is "safe." Therefore, DNS queries could be exposed and leak private data. An attacker on the same wireless network could see DNS queries, determine who your employer is, or even intercept them and [respond with malicious answers](https://github.com/iphelix/dnschef). Using DoH protects the DNS queries from eavesdropping and tampering.
```

* We use Cloudflare's DoH servers with basic malware blocking. No censorship should be applied on these servers, except towards destinations known as malware related.

**User experience impact**

> * Misconfigured captive portals, such as in hotels, could be unreachable, as they perform an attack on DNS traffic to redirect you, and misuse IP addresses such as *1.1.1.1*. You can work around this by performing a *nslookup* in the terminal for any domain and manually browsing to the IP being provided as the response. **This should be rare, and getting rarer by the day. The best workaround is to use tethering on a phone or hotspot**.

Reviewer: This also sounds scary and is too technical for me to suggest anything friendlier 🤯 I'm open to suggestions.

Author: Some misconfigured "captive portals", typically used in hotels and airports, might be unusable with DoH due to how they are configured. This can be worked around by using the hotspot on your phone, and if you really have to use this network for an extended period of time, there are usually workarounds that can be performed to connect to them. Navigating to http://1.1.1.1 often resolves the issue.

Author: Applying that

> * Rare false positives could happen, preventing access to a site. Please report those, as if they become frequent, we will define a strategy for handling them.

Reviewer: This sounds like something I should be aware of as an employee, but I don't understand it. Please consider rephrasing it for the layman.

Author: What about: If you are trying to reach a site, and you believe it is being blocked accidentally, please submit it to [Cloudflare](https://radar.cloudflare.com/categorization-feedback/). This should be extremely rare. If it is not, please let the security team know.

Author: Applying this

> * If your ISP's DNS service goes down, you'll be able to continue working 😎

Reviewer: I like the emoji. Feel free to include them in other places. They make me feel safe 😄

> *Note: Implementing DoH in an office environment where other network controls are in place has different downsides than doing it for a remote company. In those cases, **disabling** DoH makes more sense, so network controls can retain visibility. Please evaluate your situation before implementing any of our recommendations at your organization, especially DoH.*

Reviewer: As we are an all-remote company, I don't think this not is necessary?

Author: This is for companies who want to base their own policy on ours - maybe we should add a note before that? Or completely remove it?

#### Deploy osquery
| #  | Setting                |
| -- | ---------------------- |
| NA | Deploy [osquery](https://osquery.io/) pointed to our dogfood instance |

***Why?***

> We leverage osquery and Fleet to monitor our own devices. This is used for vulnerability detection, security posture tracking, and can be used for incident response when necessary.

Reviewer: ```suggestion
We use osquery and Fleet to monitor our own devices. This is used for vulnerability detection, security posture tracking, and can be used for incident response when necessary.
```

### Chrome configuration

> We configure Chrome on company owned devices with a basic policy.

Reviewer: ```suggestion
We configure Chrome on company-owned devices with a basic policy.
```

| Setting                                                   |
| --------------------------------------------------------- |
| Enforce Chrome updates and Chrome restart within 48 hours |
| Block intrusive ads                                       |
| uBlock Origin ad blocker extension deployed               |
| Password manager extension deployed                       |
| Chrome Endpoint Verification extension deployed           |

**Why?**

> * Browsers have large attack surface, and their updates contain critical security updates. 

Reviewer: ```suggestion
* Browsers have a large attack surface, and their updates contain critical security updates. 
```

**User experience impact**

> * Chrome will need to be restarted within 48 hours of patches being installed. The automatic restart happens after 19:00 and before 6:00 if the computer is running, and tabs are restored (except for incognito tabs).

Reviewer: ```suggestion
* Chrome needs to be restarted within 48 hours of patches being installed. The automatic restart happens after 19:00 and before 6:00 if the computer is running, and tabs are restored (except for incognito tabs).
```

* Ads considered intrusive are blocked.
* uBlock Origin is enabled by default, and is 100% configurable, improving security and performance of browsing.

> * Endpoint Verification is used to make access decision based on the security posture of the computer. For example, an outdated Mac could be prevented access to Google Drive.

Reviewer: ```suggestion
* Endpoint Verification is used to make access decisions based on the security posture of the device. For example, an outdated Mac could be prevented access to Google Drive.
```

**Why?**

* Fleet employees are free to work from wherever they choose. If a laptop is lost or forgotten, automatic login exposes sensitive company data and poses a critical security risk. 
* Password hints can sometimes be easier to guess than the password itself. Since we support employees remotely via MDM and do not require users to change passwords frequently, we eliminate the need for passwords hints and their associated risk.
* Only a single employee should use a company laptop. Therefore, guest accounts are not permitted.
* Automatic logon would defeat the purpose of even requiring passwords to unlock computers.

**User experience impact**

* Laptops lock after 20 minutes of inactivity. To voluntarily pause this, a [hot corner](https://support.apple.com/en-mo/guide/mac-help/mchlp3000/mac) can be configured to disable the screen saver. This is useful if you are, for example, watching an online meeting without moving the mouse and want to be sure the laptop will not lock.
* Forgotten passwords can be fixed via MDM, instead of relying on potentially dangerous hints.
* Guest accounts are not available.

#### iCloud

> We do not apply ultra restrictive Data Loss Prevention style policies to our devices. Instead, by using different web-based tools we have, we ensure that the most critical company data is not stored on our laptops, so it can remain secure, while our laptops can remain productive.

Reviewer: ```suggestion
We do not apply ultra restrictive Data Loss Prevention style policies to our devices. Instead, by using our company Google Drive, we ensure that the most critical company data never reaches our laptops, so it can remain secure, while our laptops can remain productive.
```

**Why?**
* We do not use managed Apple IDs, and allow employees to use their own iCloud accounts. We disable iCloud Documents and Desktop sync to avoid "accidental" copying of data to iCloud, but we do allow iCloud drive.

**User experience impact**

* iCloud remains allowed, but the Desktop and Documents folders will not be synchronized. Ensure you put your documents in our Google Drive, so you do not lose them if your laptop has an issue.

#### Miscellaneous security settings

| #     | Setting                                                      |
| ----- | ------------------------------------------------------------ |
| 2.5.6 | Ensure limit ad tracking is enabled                          |
| 2.10  | Ensure secure keyboard entry Terminal.app is enabled         |
| 5.1.4 | Ensure library validation is enabled                         |
| 6.3   | Ensure automatic opening of safe files in Safari is disabled |

**Why?**

* Limiting ad tracking has privacy benefits, and no downside.
* Protecting keyboard entry into Terminal.app could prevent malicious applications or non-malicious but inappropriate applications from receiving passwords.

> * Library validation ensures that an attacker can't trick applications into loading a software library in a different location, leaving it open to abuse.

Reviewer: ```suggestion
* Library validation ensures that an attacker can't trick applications into loading a software library in a different location, leaving it open to abuse.
```
